crates/iceberg/src/spec/values.rs

```diff
@@ -0,0 +1,961 @@
+                        Ok(Literal::Primitive(PrimitiveLiteral::Boolean(false)))
+                    } else {
+                        Ok(Literal::Primitive(PrimitiveLiteral::Boolean(true)))
+                    }
+                }
+                PrimitiveType::Int => Ok(Literal::Primitive(PrimitiveLiteral::Int(
+                    i32::from_le_bytes(bytes.try_into()?),
+                ))),
+                PrimitiveType::Long => Ok(Literal::Primitive(PrimitiveLiteral::Long(
+                    i64::from_le_bytes(bytes.try_into()?),
+                ))),
+                PrimitiveType::Float => Ok(Literal::Primitive(PrimitiveLiteral::Float(
+                    OrderedFloat(f32::from_le_bytes(bytes.try_into()?)),
+                ))),
+                PrimitiveType::Double => Ok(Literal::Primitive(PrimitiveLiteral::Double(
+                    OrderedFloat(f64::from_le_bytes(bytes.try_into()?)),
+                ))),
+                PrimitiveType::Date => Ok(Literal::Primitive(PrimitiveLiteral::Date(
+                    date::days_to_date(i32::from_le_bytes(bytes.try_into()?))?,
+                ))),
+                PrimitiveType::Time => Ok(Literal::Primitive(PrimitiveLiteral::Time(
+                    time::microseconds_to_time(i64::from_le_bytes(bytes.try_into()?))?,
+                ))),
+                PrimitiveType::Timestamp => Ok(Literal::Primitive(PrimitiveLiteral::Timestamp(
+                    timestamp::microseconds_to_datetime(i64::from_le_bytes(bytes.try_into()?))?,
+                ))),
+                PrimitiveType::Timestamptz => Ok(Literal::Primitive(
+                    PrimitiveLiteral::TimestampTZ(timestamptz::microseconds_to_datetimetz(
+                        i64::from_le_bytes(bytes.try_into()?),
+                    )?),
+                )),
+                PrimitiveType::String => Ok(Literal::Primitive(PrimitiveLiteral::String(
+                    std::str::from_utf8(bytes)?.to_string(),
+                ))),
+                PrimitiveType::Uuid => Ok(Literal::Primitive(PrimitiveLiteral::UUID(
+                    Uuid::from_u128(u128::from_be_bytes(bytes.try_into()?)),
+                ))),
+                PrimitiveType::Fixed(_) => Ok(Literal::Primitive(PrimitiveLiteral::Fixed(
+                    Vec::from(bytes),
+                ))),
+                PrimitiveType::Binary => Ok(Literal::Primitive(PrimitiveLiteral::Binary(
+                    Vec::from(bytes),
+                ))),
+                PrimitiveType::Decimal {
+                    precision: _,
+                    scale: _,
+                } => todo!(),
+            },
+            _ => Err(Error::new(
+                crate::ErrorKind::DataInvalid,
+                "Converting bytes to non-primitive types is not supported.",
+            )),
+        }
+    }
+
+    /// Create iceberg value from a json value
+    pub fn try_from_json(value: JsonValue, data_type: &Type) -> Result<Self, Error> {
```

Comment:
Do we need consider the JsonValue::Null case?

Reply:
I commit some changes to solve this. Let me know what you think.

Reply:
Actually my solution is to change the interface like this:

```
pub fn try_from_json(value: JsonValue, data_type: &Type) -> Result<Option<Self>, Error> {
   if value.is_null() {
       return Ok(None)
   }
    ...
    match data_type {
        match DataType::List =>{
            array
                .into_iter()
                .map(|value| {
                    (Literal::try_from_json(
                        value,
                        &list.element_field.field_type,
                    ))
                })
        }
    }
}
```

So that we don't need `to_optional_literal` to check result.

Reply:
From a user perspective I prefer the current function signature `pub fn try_from_json(value: JsonValue, data_type: &Type) -> Result<Self, Error>` because the iceberg spec doesn't specify optional values(null) at the top level. The optional value would then need to be handled elsewhere.

Reply:
Yes, I also think about this. That's why we seperate Literal and Option<Literal>. None is not a value in Literal. 

And IMO I think following signature will not change this semantics. 
```
// Convert a json value into a optional literal value. It didn't change the semantics of literal.
try_from_json(value: JsonValue, data_type: &Type) -> Result<Option<Self>, Error>
```

Or can we maintain two version, I'm not sure whether is necessary, but it seems more clean.
`try_from_json(value: JsonValue, data_type: &Type) -> Result<Self, Error>`
`try_from_json_optional(value: JsonValue, data_type: &Type) -> Result<Option<Self>, Error>`

Reply:
Sorry for missing this discussion: https://github.com/apache/iceberg-rust/pull/20#pullrequestreview-1568938352

Reply:
Ideally, `try_from_json` should not be called by our users, it's our internal logic.